logdna/logdna.py

```diff
@@ -135,28 +141,34 @@ def flush(self):
             except Exception as e:
                 self.internalLogger.debug('Error in calling flush: %s', e)
 
-    def flush_sync(self):
-        if self.buf_size == 0 and len(self.secondary) == 0:
-            return
+    def flush_sync(self, should_block=False):
```

Comment:
"sync" in name implies blocking? isn't? 
why it is optional now?

Reply:
Previously, it was never blocking. The reason it is blocking now is because it is actually invoked in two contexts, one of which was previously subject to a race:
* When the flusher Timer thread wakes up
* When the size of the buffered data meets or exceeds the configured buffer size send threshold
* When the logger is being closed. This is where a race was identified where the request threads would be shut down while the worker threads were adding data to the buffer.  In this case, we want the final shutdown flush to block and wait to acquire the buffer lock so we can be sure to do a final drain and not lose any logged data.

Happy to consider renaming this method if anyone has any better suggestions.

Reply:
if flush_sync() was not blocking then it is a bug.

Reply:
I actually don't think it needs to (or should) be blocking in all circumstances. Because the method can get invoked in separate threads, there's not really a benefit to one thread waiting to acquire the lock when another is already active trying to drain the buffer. Whoever gets the lock first can complete the sync and re-initialize the buffer. The only reason we want it to block when waiting to acquire the lock during shutdown is to avoid a race where the lock isn't released quickly enough by other threads during shutdown.

Reply:
so the fix is for app shutdown case only?
what is the meaning of the "sync" here?

Reply:
@dkhokhlov More specifically, there were a few different races that were identified and addressed with this PR. The change to make the behavior here blocking for this specific scenario addresses one of those races.

As far as the term "sync", I can see how sync would imply "synchronized" or "synchronous", but I actually interpreted it differently when I first came across it and read the code, since it's intended purposes is to "sync" the data from the client to the server.  I'm not opinionated here on naming though and am happy to change it if you and others feel strongly, but I think that the actual meat and potatoes of this method is now doing what (in theory) it should be.

Reply:
> few different races 
the test is for shutdown only?

Reply:
The test covers all of the potential race scenarios that I uncovered. The main race is exposed when the logger is invoked at a rate sufficiently fast enough (which is not all that fast in actuality). In that scenario, the buffer would be drained while it was being written to by other threads. The other race is that one we were discussing previously during shutdown. The test covers the former case by sending a high volume of logs (10k) as quickly as possible, stressing the access synchronization to the buffer, and covers the latter case by shutting down the logger at the end and then subsequently verifying that all expected data was emitted by the logger.